### DISCUSSION TITLE
Fix `for` 0..n-1 iteration to `foreach`

### DISCUSSION THREAD

**Discussion by mvorisek** [PHP-CS-Fixer/PHP-CS-Fixer#7286] [Feature ideas]
## Feature request

In legacy codes I find very often code patterns like:
```
for ($i = 0; $i < count($arr); $i++) {
    $this->doSomething($arr[$i]);
}
```
which should be converted to:
```
foreach ($arr as $i => $v) {
    $this->doSomething($v);
}
```
the later code is much more readable and faster.

`$i => ` part can be removed if `$i` is not used in the script further.

**Comment by TimWolla** [comment]
> the later code is much more readable and faster.

It's also not equivalent:

```
<?php
$arr = [1, 2, 3, 4, 5, 6];

for ($i = 0; $i < count($arr); $i++) {
    unset($arr[$i - 1]);
    var_dump($arr[$i]);
}
echo "===", PHP_EOL;
$arr = [1, 2, 3, 4, 5, 6];

foreach ($arr as $i => $v) {
    unset($arr[$i - 1]);
    var_dump($arr[$i]);
}
```

https://3v4l.org/bQGE3

**Comment by Wirone** [comment]
After @TimWolla's comment and additional thinking about it, I am not sure if it's worth the hassle. Fixer is not a complete SCA tool and lack of code context may be a problem here. Such fixer would be risky by definition and since we don't have mechanism for ignoring parts of the code, it could be problematic for some codebases and end up as disabled via config.

With IDE regex search and replace you can use `for\s*\((\$[a-zA-Z0-9]+)\s*=\s*[0-9]+;\s*\$[a-zA-Z0-9]+\s*[<>]\s*count\((\$[a-zA-Z0-9]+)\);\s*\$[a-zA-Z0-9]+[+-]*\)\s*{?` and replace occurrences with `foreach`:

![image](https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/assets/600668/30e7702c-9d1f-4dde-b6c4-912cad8034bb)

Things that should be taken into consideration within such fixer:
- what happens if there's `$varCount` instead of `count($var)` (look for the variable assignment? ignore?)
- variable representing value introduced in `foreach` may collide with the current code and there's no way to figure it out easily
- as pointed above, what's inside `for` block matters, so such change can't be done blindly

That's why I think we should close this feature request, especially that it did not get attention from other people. But I am going to convert this into discussion, because it should be created there in the first place.
